ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineContext.java

```diff
@@ -208,16 +231,90 @@ DdlCommand createDdlCommand(
   String executeDdl(
       final String sqlExpression,
       final DdlCommand command,
-      final boolean withQuery
+      final boolean withQuery,
+      final Set<SourceName> withQuerySources
   ) {
-    final DdlCommandResult result = ddlCommandExec.execute(sqlExpression, command, withQuery);
+    if (command instanceof DropSourceCommand) {
+      throwIfOtherQueriesExist(((DropSourceCommand) command).getSourceName());
+    }
+
+    final DdlCommandResult result =
+        ddlCommandExec.execute(sqlExpression, command, withQuery, withQuerySources);
     if (!result.isSuccess()) {
       throw new KsqlStatementException(result.getMessage(), sqlExpression);
     }
+
+    if (command instanceof DropSourceCommand) {
+      terminateCreateAsQuery(((DropSourceCommand) command).getSourceName());
```

Comment:
can this fail for any reason? we might be left in a state where we drop the source but still have a query running into it. That would be very bad 😢 is there any way to make the drop & terminate an atomic operation?

Reply:
It's hard to make this atomic. I would need to change the code to mark the Metastore source as deleted, but not real deletion. Then attempt to terminate the query, if successful, then do the real delete, otherwise undo the deletion mark. 

Another solution I thought was to attempt to terminate the query before deleting the source, but there is another referential check inside the Metastore that should be done before terminating the query, so I couldn't either.

Anyway, seems the only reason the `query.close()` would fail is if the `closeTimeout` is invalid. I made a change in the `QueryMetadata` to build the `Duration.fromMillis` during the constructor. That way we are sure the close won't fail because of that. And the other reasons that could make this to fail are thrown after the query was marked for `PENDING_SHUTDOWN`. After this is set, the stream cannot be closed again 'cause it will still pending shutting down. Not sure what else to do there.